3-async-await/solution.js

```diff
@@ -20,16 +23,151 @@ Example:
 9. as extra challenge: add Promise.race() and Promise.any(), and try to get the idea of what happens
 */
 
-function solution() {
-    // YOUR SOLUTION GOES HERE
+class Product {
+    constructor(id, name, price){
+        this.id = id
+        this.name = name
+        this.price = price
+    }
+    print(){
+        console.log(`${'*'.repeat(30)}`)
+        console.log(`ID: ${this.id} | product: ${this.name} | price: ${this.price.toFixed(2)} $`)
+        console.log(`${'*'.repeat(30)}`)
+    }
+}
+
+class CustomError extends Error{
+    constructor(message, origin, id){
+        super(message)
+        this.origin = origin
+        this.id = id
+    }
+    print(){
+        console.error(`${'='.repeat(30)}`)
+        console.error(`**ERROR OCCURRED`)
+        console.error(`- ID: ${this.id}`)
+        console.error(`- Origin: ${this.origin || 'unknown'}`)
+        console.error(`- Error Message: ${this.message}`)
+        console.error(`${'='.repeat(30)}`)
+    }
+}
+
+class Result {
+    constructor(product, error, timelapse){
+        this.product = product
+        this.error = error
+        this.timelapse = timelapse
+    }
+    setProduct(product){
+        this.product = product
+    }
+    setError(error){
+        this.error = error
+    }
+    setTimelapse(timelapse){
+        this.timelapse = timelapse
+    }
+
+    print(){
+        console.log(`\nResult in: ${this.timelapse} miliseconds`)
+        if(this.error){
+            this.error.print()
+        }else{
+            this.product.print()
+        }
+    }
+}
+
+const generateId = async(delay) => {
+    if(delay){
+        await setTimeout(100)
```

Comment:
You can use the delay parameter

Reply:
That was the idea originally, but I either forgot or accidentally changed it during some testing. It has already been resolved in this commit: https://github.com/ravnhq/JS-async-Nerdery-Challenges/pull/33/commits/058850dcf79252d35285d183d38cf05091aab519